src/gui/tlserrordialog.cpp

```diff
@@ -0,0 +1,93 @@
+#include "tlserrordialog.h"
+#include "common/utility.h"
+#include "ui_tlserrordialog.h"
+
+#include <QAbstractButton>
+#include <QPushButton>
+
+namespace OCC {
+TlsErrorDialog::TlsErrorDialog(const QList<QSslError> &sslErrors, const QString &host, QWidget *parent)
+    : QDialog(parent)
+    , _ui(new Ui::TlsErrorDialog)
+{
+    _ui->setupUi(this);
+
+    _ui->hostnameLabel->setText(tr("Cannot connect securely to %1").arg(host));
+
+    QStringList errorStrings;
+
+    for (const auto &error : sslErrors) {
+        errorStrings << error.errorString() << describeCertificateHtml(error.certificate());
+    }
+
+    _ui->textBrowser->setHtml(errorStrings.join("\n"));
+
+    // FIXME: add checkbox for second confirmation
+
+    connect(_ui->buttonBox, &QDialogButtonBox::accepted, this, [this]() {
+        qDebug() << "accepted";
+        accept();
+    });
+    connect(_ui->buttonBox, &QDialogButtonBox::rejected, this, [this]() {
+        qDebug() << "rejected";
+        reject();
+    });
+
+    // of course, we require an answer from the user, they may not proceed with anything else
+    setModal(true);
+}
+
+TlsErrorDialog::~TlsErrorDialog()
+{
+    delete _ui;
+}
+
+QString TlsErrorDialog::describeCertificateHtml(const QSslCertificate &certificate)
+{
+    QString msg;
+    msg += QStringLiteral("<div id=\"cert\">");
```

Comment:
Don't append multiple times, with msg +=;
This breaks the benefits of the used stringbuilder.

Reply:
This was copied 1:1 from the old dialog. I'd rather not touch this at this point, but we could leave a TODO to improve it in the future (when removing the SSL error dialog).

Reply:
Best time to refactor code is when you touch it.